ts/src/graphql/query/edge_connection.ts

```diff
@@ -56,16 +57,37 @@ export class GraphQLEdgeConnection<
     args?: Data,
   ) {
     this.viewer = viewer;
+
+    async function resolveQuery(
+      query: MaybePromise<EdgeQuery<TSource, Ent, TEdge>>,
+      source: TSource | undefined,
+    ) {
+      const resolved = await query;
+
+      // https://github.com/lolopinto/ent/issues/1836
+      if (resolved instanceof CustomClauseQuery && source !== undefined) {
+        resolved.__setSource(source);
+      }
+      return resolved;
+    }
+
+    let maybeQuery: MaybePromise<EdgeQuery<TSource, Ent, TEdge>>;
+    let source: TSource | undefined;
+
     if (typeof arg2 === "function") {
-      this.query = Promise.resolve(arg2(this.viewer));
+      maybeQuery = arg2(this.viewer);
     } else {
       this.source = arg2;
+      source = arg2;
```

Comment:
Do you need this extra variable? Could it just be `this.source` that you pass in?

Reply:
Oh I see, playing it safe for the async function?

Reply:
It wasn't resolving. I can try harder to figure out why lol

Reply:
Up to you. There are also some `this.source!` uses here too, which gave me a head tilt.

Reply:
this doesn't resolve because JS closures